bodhi/server/models.py

```diff
@@ -2234,6 +2234,13 @@ def _get_test_gating_status(self):
         for data in self.greenwave_request_batches(verbose=False):
             response = util.greenwave_api_post(self._greenwave_api_url, data)
             if not response['policies_satisfied']:
+                # if there are no failures (only missing results) and
+                # it's less than two hours since the update was modified,
+                # let's assume tests are in progress
+                unsats = response.get('unsatisfied_requirements', [])
+                if unsats and all(req.get('type', '') == 'test-result-missing' for req in unsats):
+                    if datetime.utcnow() - self.last_modified < timedelta(hours=2):
+                        return TestGatingStatus.waiting
```

Comment:
The whole loop is a bit complex and hard to reason about.

Maybe replace the loop with:

```python
responses = [
    util.greenwave_api_post(self._greenwave_api_url, data)
    for data in self.greenwave_request_batches(verbose=False)
]

if all(r['summary'] != 'no tests are required' for r in responses):
    return TestGatingStatus.ignored

if all(r['policies_satisfied'] for r in responses):
    return TestGatingStatus.passed

# if there are no failures (only missing results) and
# it's less than two hours since the update was modified,
# let's assume tests are in progress
if datetime.utcnow() - self.last_modified < timedelta(hours=2):
    unsats = []
    for response in responses:
        unsats.extend(response.get('unsatisfied_requirements', []))
    if unsats and all(req.get('type', '') == 'test-result-missing' for req in unsats):
        return TestGatingStatus.waiting

return TestGatingStatus.failed
```

Reply:
I agree the loop is kind of a mess, and actually, now I look at both versions a bit longer, mine is wrong - we can't safely just `return TestGatingStatus.waiting` if we hit one 'waiting' case like we can safely return `TestGatingStatus.failed` if we hit one 'failed' case, because 'waiting' isn't the worst possible outcome. My version would do the wrong thing if e,g, the first response encountered had only 'not yet present' results, but the *second* response encountered had a 'failed' result. So it clearly *is* hard to reason about because I didn't reason about it right :D

I'll stare at your version for a bit and see if I see anything that needs tweaking, then replace it, thanks.

Reply:
Well, I came up with what I think is an even better option.